From 1b12ef3a01f1cd20c3126a084d006bb457407205 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Thu, 23 Jul 2026 04:14:22 -0400
Subject: [PATCH 220/268] atidxx: Implement deferred contexts.

so that's what those weird offsets did all along huh
---
 dlls/atidxx64/atidxx.h |   9 +-
 dlls/atidxx64/main.c   | 243 +++++++++++++++++++++++++++++++++--------
 2 files changed, 202 insertions(+), 50 deletions(-)

diff --git a/dlls/atidxx64/atidxx.h b/dlls/atidxx64/atidxx.h
index 3785a3c7e34..d963bc7f280 100644
--- a/dlls/atidxx64/atidxx.h
+++ b/dlls/atidxx64/atidxx.h
@@ -64,8 +64,6 @@ DECLARE_INTERFACE_(IAmdDxExt, IAmdDxExtInterface)
     /*** IAmdDxExt methods ***/
     THISCALLMETHOD_(HRESULT, GetVersion)(THIS_ AmdDxExtVersion *version) PURE;
     THISCALLMETHOD_(IAmdDxExtInterface*,GetExtInterface)(THIS_ unsigned int iface) PURE;
-
-
     THISCALLMETHOD_(HRESULT, IaSetPrimitiveTopology)(THIS_ AmdDxExtPrimitiveTopology topology) PURE;
     THISCALLMETHOD_(HRESULT, IaGetPrimitiveTopology)(THIS_ AmdDxExtPrimitiveTopology *topology) PURE;
     THISCALLMETHOD_(HRESULT, SetSingleSampleRead)(THIS_ ID3D10Resource *res, BOOL single_sample) PURE;
@@ -88,6 +86,8 @@ DECLARE_INTERFACE_(IAmdDxExtUAVOverlap, IAmdDxExtInterface)
     THISCALLMETHOD_(HRESULT, BeginUAVOverlap)(THIS) PURE;
     THISCALLMETHOD_(HRESULT, EndUAVOverlap)(THIS) PURE;
     THISCALLMETHOD_(void, GetVersion)(THIS, AmdDxExtVersion* version) PURE;
+    THISCALLMETHOD_(HRESULT, BeginUAVOverlapCtx)(THIS, ID3D11DeviceContext *ctx) PURE;
+    THISCALLMETHOD_(HRESULT, EndUAVOverlapCtx)(THIS, ID3D11DeviceContext *ctx) PURE;
 };
 #undef INTERFACE
 
@@ -117,6 +117,7 @@ DECLARE_INTERFACE_(IAmdDxExtDepthBounds, IAmdDxExtInterface)
     /*** IAmdDxExtDepthBounds methods ***/
     THISCALLMETHOD_(HRESULT, SetDepthBounds)(THIS, BOOL enabled, float min, float max) PURE;
     THISCALLMETHOD_(void, GetVersion)(THIS, AmdDxExtVersion *version) PURE;
+    THISCALLMETHOD_(HRESULT, SetDepthBoundsCtx)(THIS, BOOL enabled, float min, float max, ID3D11DeviceContext *context) PURE;
 };
 #undef INTERFACE
 
@@ -134,6 +135,10 @@ DECLARE_INTERFACE_(IAmdDxExtMultidrawIndirect, IAmdDxExtInterface)
     THISCALLMETHOD_(HRESULT, MultiDrawIndexedIndirect)(THIS, unsigned int draw_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride) PURE;
     THISCALLMETHOD_(HRESULT, MultiDrawIndirectCount)(THIS, ID3D11Buffer *buffer_for_count, unsigned int byte_offset_for_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride) PURE;
     THISCALLMETHOD_(HRESULT, MultiDrawIndexedIndirectCount)(THIS, ID3D11Buffer *buffer_for_count, unsigned int byte_offset_for_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride) PURE;
+    THISCALLMETHOD_(HRESULT, MultiDrawIndirectCtx)(THIS, unsigned int draw_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride, ID3D11DeviceContext *ctx) PURE;
+    THISCALLMETHOD_(HRESULT, MultiDrawIndexedIndirectCtx)(THIS, unsigned int draw_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride, ID3D11DeviceContext *ctx) PURE;
+    THISCALLMETHOD_(HRESULT, MultiDrawIndirectCountCtx)(THIS, ID3D11Buffer *buffer_for_count, unsigned int byte_offset_for_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride, ID3D11DeviceContext *ctx) PURE;
+    THISCALLMETHOD_(HRESULT, MultiDrawIndexedIndirectCountCtx)(THIS, ID3D11Buffer *buffer_for_count, unsigned int byte_offset_for_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride, ID3D11DeviceContext *ctx) PURE;
 };
 
 #ifdef __cplusplus
diff --git a/dlls/atidxx64/main.c b/dlls/atidxx64/main.c
index 88f8d9bee63..af10bc6e623 100644
--- a/dlls/atidxx64/main.c
+++ b/dlls/atidxx64/main.c
@@ -127,7 +127,8 @@ HRESULT WINAPI AmdD3D11CreateDeviceAndSwapChainExt(IDXGIAdapter *adapter, D3D_DR
     0x18 = BeginUAVOverlap
     0x20 = EndUAVOverlap
     0x28 = GetVersion (called on init, prob some kind of version getter)
-
+    0x30 = BeginUAVOverlapCtx
+    0x38 = EndUAVOverlapCtx
 */
 
 /* field_0x168 (0xb)
@@ -136,6 +137,7 @@ HRESULT WINAPI AmdD3D11CreateDeviceAndSwapChainExt(IDXGIAdapter *adapter, D3D_DR
     0x10 = virtual destructor
     0x18 = SetDepthBounds
     0x20 = GetVersion
+    0x28 = SetDepthBoundsCtx
 */
 
 /* field_0x170 (0x11)
@@ -147,6 +149,10 @@ HRESULT WINAPI AmdD3D11CreateDeviceAndSwapChainExt(IDXGIAdapter *adapter, D3D_DR
     0x28 = MultiDrawIndexedIndirect
     0x30 = MultiDrawIndirectCount
     0x38 = MultiDrawIndexedIndirectCount
+    0x40 = MultiDrawIndirectCtx
+    0x48 = MultiDrawIndexedIndirectCtx
+    0x50 = MultiDrawIndirectCountCtx
+    0x58 = MultiDrawIndexedIndirectCountCtx
 */
 
 typedef struct
@@ -157,7 +163,7 @@ typedef struct
     IAmdDxExtDepthBounds IAmdDxExtDepthBounds_iface;
     IAmdDxExtMultidrawIndirect IAmdDxExtMultidrawIndirect_iface;
     LONG ref;
-    //could also be a d3d10 device, just cast
+    /* could also be a d3d10 device, just cast (we don't implement that yet) */
     ID3D11Device *device;
     ID3D11DeviceContext *context;
     ID3D11VkExtContext *ext_context;
@@ -227,10 +233,15 @@ void __thiscall AmdDxExt_Destroy(IAmdDxExt *iface)
 DEFINE_THISCALL_WRAPPER(AmdDxExt_GetVersion, 8)
 HRESULT __thiscall AmdDxExt_GetVersion(IAmdDxExt *ext, AmdDxExtVersion *version)
 {
-    FIXME("%p %p\n", ext, version);
+    TRACE("%p %p\n", ext, version);
 
-    version->majorVersion = 1;
-    version->minorVersion = 0;
+    /* v = (version.major << 10 | version.minor) << 12
+     * if v > 0x2400fff then "deferred contexts" are allowed.
+     * With DXVK deferred context is always supported so we chilling. */
+
+    /* 0x2401000 */
+    version->majorVersion = 9;
+    version->minorVersion = 1;
 
     return S_OK;
 }
@@ -244,23 +255,24 @@ IAmdDxExtInterface* __thiscall AmdDxExt_GetExtInterface(IAmdDxExt *ext, unsigned
 
     switch (iface)
     {
-        case 0x2:
-            ret = (IAmdDxExtInterface *)&This->IAmdDxExtQuadBufferStereo_iface;
-            break;
-        case 0xb:
+    case 0x2:
+        ret = (IAmdDxExtInterface *)&This->IAmdDxExtQuadBufferStereo_iface;
+        break;
+    case 0xb:
+        if (This->depth_bounds)
             ret = (IAmdDxExtInterface *)&This->IAmdDxExtDepthBounds_iface;
-            break;
-        case 0xf:
+        break;
+    case 0xf:
+        if (This->uav_overlap)
             ret = (IAmdDxExtInterface *)&This->IAmdDxExtUAVOverlap_iface;
-            break;
-        case 0x11:
+        break;
+    case 0x11:
+        if (This->multi_draw_indirect)
             ret = (IAmdDxExtInterface *)&This->IAmdDxExtMultidrawIndirect_iface;
-            break;
-        default:
-        {
-            FIXME("Unknown interface %x\n", iface);
-            break;
-        }
+        break;
+    default:
+        FIXME("Unknown interface %x\n", iface);
+        break;
     }
 
     if (ret) AmdDxExt_AddRef(ext);
@@ -426,8 +438,6 @@ HRESULT __thiscall AmdDxExtUAVOverlap_BeginUAVOverlap(IAmdDxExtUAVOverlap *iface
 
     if (!This->ext_context) return E_FAIL;
 
-    if (!This->uav_overlap) return E_FAIL;
-
     ID3D11VkExtContext_SetBarrierControl(This->ext_context, D3D11_VK_BARRIER_CONTROL_IGNORE_WRITE_AFTER_WRITE);
 
     return S_OK;
@@ -441,19 +451,58 @@ HRESULT __thiscall AmdDxExtUAVOverlap_EndUAVOverlap(IAmdDxExtUAVOverlap *iface)
 
     if (!This->ext_context) return E_FAIL;
 
-    if (!This->uav_overlap) return E_FAIL;
-
     ID3D11VkExtContext_SetBarrierControl(This->ext_context, 0);
 
     return S_OK;
 }
 
 DEFINE_THISCALL_WRAPPER(AmdDxExtUAVOverlap_GetVersion, 8)
-void __thiscall AmdDxExtUAVOverlap_GetVersion(IAmdDxExtUAVOverlap *iface, AmdDxExtVersion* version)
+void __thiscall AmdDxExtUAVOverlap_GetVersion(IAmdDxExtUAVOverlap *iface, AmdDxExtVersion *version)
 {
-    FIXME("%p %p stub!\n", iface, version);
+    TRACE("%p %p\n", iface, version);
+
+    /* ((version.major << 10) | version.minor) > 0x403
+     * means that depth bounds deferred contexts are supported */
+
+    /* 0x404 */
     version->majorVersion = 1;
-    version->minorVersion = 0;
+    version->minorVersion = 4;
+}
+
+DEFINE_THISCALL_WRAPPER(AmdDxExtUAVOverlap_BeginUAVOverlapCtx, 8)
+HRESULT __thiscall AmdDxExtUAVOverlap_BeginUAVOverlapCtx(IAmdDxExtUAVOverlap *iface, ID3D11DeviceContext *ctx)
+{
+    ID3D11VkExtContext *ext_context;
+
+    TRACE("%p %p\n", iface, ctx);
+
+    if (!ctx) return E_FAIL;
+
+    if (FAILED(ID3D11DeviceContext_QueryInterface(ctx, &IID_ID3D11VkExtContext, (void **)&ext_context)))
+        return E_FAIL;
+
+    ID3D11VkExtContext_SetBarrierControl(ext_context, D3D11_VK_BARRIER_CONTROL_IGNORE_WRITE_AFTER_WRITE);
+    ID3D11VkExtContext_Release(ext_context);
+
+    return S_OK;
+}
+
+DEFINE_THISCALL_WRAPPER(AmdDxExtUAVOverlap_EndUAVOverlapCtx, 8)
+HRESULT __thiscall AmdDxExtUAVOverlap_EndUAVOverlapCtx(IAmdDxExtUAVOverlap *iface, ID3D11DeviceContext *ctx)
+{
+    ID3D11VkExtContext *ext_context;
+
+    TRACE("%p %p\n", iface, ctx);
+
+    if (!ctx) return E_FAIL;
+
+    if (FAILED(ID3D11DeviceContext_QueryInterface(ctx, &IID_ID3D11VkExtContext, (void **)&ext_context)))
+        return E_FAIL;
+
+    ID3D11VkExtContext_SetBarrierControl(ext_context, 0);
+    ID3D11VkExtContext_Release(ext_context);
+
+    return S_OK;
 }
 
 DEFINE_THISCALL_WRAPPER(AmdDxExtQuadBufferStereo_AddRef, 4)
@@ -521,9 +570,14 @@ unsigned int __thiscall AmdDxExtDepthBounds_Release(IAmdDxExtDepthBounds *iface)
 DEFINE_THISCALL_WRAPPER(AmdDxExtDepthBounds_GetVersion, 8)
 void __thiscall AmdDxExtDepthBounds_GetVersion(IAmdDxExtDepthBounds *iface, AmdDxExtVersion* version)
 {
-    FIXME("%p %p stub!\n", iface, version);
+    TRACE("%p %p\n", iface, version);
+
+    /* ((version.major << 10) | version.minor) > 0x400
+     * means that depth bounds deferred contexts are supported */
+
+    /* 0x401 */
     version->majorVersion = 1;
-    version->minorVersion = 0;
+    version->minorVersion = 1;
 }
 
 DEFINE_THISCALL_WRAPPER(AmdDxExtDepthBounds_SetDepthBounds, 16)
@@ -535,13 +589,29 @@ HRESULT __thiscall AmdDxExtDepthBounds_SetDepthBounds(IAmdDxExtDepthBounds *ifac
 
     if (!This->ext_context) return E_FAIL;
 
-    if (!This->depth_bounds) return E_FAIL;
-
     ID3D11VkExtContext_SetDepthBoundsTest(This->ext_context, enabled, min, max);
 
     return S_OK;
 }
 
+DEFINE_THISCALL_WRAPPER(AmdDxExtDepthBounds_SetDepthBoundsCtx, 20)
+HRESULT __thiscall AmdDxExtDepthBounds_SetDepthBoundsCtx(IAmdDxExtDepthBounds *iface, BOOL enabled, float min, float max, ID3D11DeviceContext *context)
+{
+    ID3D11VkExtContext *ext_context;
+
+    TRACE("%p %u %f %f %p\n", iface, enabled, min, max, context);
+
+    if (!context) return E_INVALIDARG;
+
+    if (FAILED(ID3D11DeviceContext_QueryInterface(context, &IID_ID3D11VkExtContext, (void **)&ext_context)))
+        return E_FAIL;
+
+    ID3D11VkExtContext_SetDepthBoundsTest(ext_context, enabled, min, max);
+    ID3D11VkExtContext_Release(ext_context);
+
+    return S_OK;
+}
+
 DEFINE_THISCALL_WRAPPER(AmdDxExtDepthBounds_Destroy, 4)
 void __thiscall AmdDxExtDepthBounds_Destroy(IAmdDxExtDepthBounds *iface)
 {
@@ -572,13 +642,13 @@ DEFINE_THISCALL_WRAPPER(AmdDxExtMultidrawIndirect_GetVersion, 8)
 void __thiscall AmdDxExtMultidrawIndirect_GetVersion(IAmdDxExtMultidrawIndirect *iface, AmdDxExtVersion* version)
 {
     AmdDxExt *This = impl_from_IAmdDxExtMultidrawIndirect(iface);
-    FIXME("%p %p semi-stub!\n", iface, version);
+    TRACE("%p %p\n", iface, version);
     /* uVar3 = (version.major << 10 | version.minor) << 12 */
     /* 0x401fff < uVar3 -> multi draw indirect count is supported */
-    /* 0x403fff < uVar3 -> different version of multidraw indirect that has different offsets (maybe for crossfire) */
+    /* 0x403fff < uVar3 -> deferred context, but implies the above */
+    /* 0x404000 */
     version->majorVersion = 1;
-    /* needed for indirect count */
-    version->minorVersion = This->multi_draw_indirect_count ? 2 : 0;
+    version->minorVersion = This->multi_draw_indirect_count ? 4 : 0;
 }
 
 DEFINE_THISCALL_WRAPPER(AmdDxExtMultidrawIndirect_MultiDrawIndirect, 20)
@@ -589,8 +659,6 @@ HRESULT __thiscall AmdDxExtMultidrawIndirect_MultiDrawIndirect(IAmdDxExtMultidra
 
     if (!This->ext_context) return E_FAIL;
 
-    if (!This->multi_draw_indirect) return E_FAIL;
-
     ID3D11VkExtContext_MultiDrawIndirect(This->ext_context, draw_count, buffer, byte_offset, byte_stride);
 
     return S_OK;
@@ -604,8 +672,6 @@ HRESULT __thiscall AmdDxExtMultidrawIndirect_MultiDrawIndexedIndirect(IAmdDxExtM
 
     if (!This->ext_context) return E_FAIL;
 
-    if (!This->multi_draw_indirect) return E_FAIL;
-
     ID3D11VkExtContext_MultiDrawIndexedIndirect(This->ext_context, draw_count, buffer, byte_offset, byte_stride);
 
     return S_OK;
@@ -635,15 +701,14 @@ static unsigned int get_max_draw_count(ID3D11Buffer *buffer, unsigned int offset
     return count;
 }
 
-DEFINE_THISCALL_WRAPPER(AmdDxextMultiDrawIndirect_MultiDrawIndirectCount, 24)
-HRESULT __thiscall AmdDxextMultiDrawIndirect_MultiDrawIndirectCount(IAmdDxExtMultidrawIndirect *iface, ID3D11Buffer *buffer_for_count, unsigned int byte_offset_for_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride)
+DEFINE_THISCALL_WRAPPER(AmdDxExtMultiDrawIndirect_MultiDrawIndirectCount, 24)
+HRESULT __thiscall AmdDxExtMultiDrawIndirect_MultiDrawIndirectCount(IAmdDxExtMultidrawIndirect *iface, ID3D11Buffer *buffer_for_count, unsigned int byte_offset_for_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride)
 {
     AmdDxExt *This = impl_from_IAmdDxExtMultidrawIndirect(iface);
     unsigned int max_draw_count;
     TRACE("%p %p %u %p %u %u\n", iface, buffer_for_count, byte_offset_for_count, buffer, byte_offset, byte_stride);
 
     if (!This->ext_context) return E_FAIL;
-
     if (!This->multi_draw_indirect_count) return E_FAIL;
 
     max_draw_count = get_max_draw_count(buffer, byte_offset, byte_stride, sizeof(D3D11_DRAW_INDEXED_INSTANCED_INDIRECT_ARGS));
@@ -653,15 +718,14 @@ HRESULT __thiscall AmdDxextMultiDrawIndirect_MultiDrawIndirectCount(IAmdDxExtMul
     return S_OK;
 }
 
-DEFINE_THISCALL_WRAPPER(AmdDxextMultiDrawIndirect_MultiDrawIndexedIndirectCount, 24)
-HRESULT __thiscall AmdDxextMultiDrawIndirect_MultiDrawIndexedIndirectCount(IAmdDxExtMultidrawIndirect *iface, ID3D11Buffer *buffer_for_count, unsigned int byte_offset_for_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride)
+DEFINE_THISCALL_WRAPPER(AmdDxExtMultiDrawIndirect_MultiDrawIndexedIndirectCount, 24)
+HRESULT __thiscall AmdDxExtMultiDrawIndirect_MultiDrawIndexedIndirectCount(IAmdDxExtMultidrawIndirect *iface, ID3D11Buffer *buffer_for_count, unsigned int byte_offset_for_count, ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride)
 {
     AmdDxExt *This = impl_from_IAmdDxExtMultidrawIndirect(iface);
     unsigned int max_draw_count;
     TRACE("%p %p %u %p %u %u\n", iface, buffer_for_count, byte_offset_for_count, buffer, byte_offset, byte_stride);
 
     if (!This->ext_context) return E_FAIL;
-
     if (!This->multi_draw_indirect_count) return E_FAIL;
 
     max_draw_count = get_max_draw_count(buffer, byte_offset, byte_stride, sizeof(D3D11_DRAW_INDEXED_INSTANCED_INDIRECT_ARGS));
@@ -671,6 +735,82 @@ HRESULT __thiscall AmdDxextMultiDrawIndirect_MultiDrawIndexedIndirectCount(IAmdD
     return S_OK;
 }
 
+DEFINE_THISCALL_WRAPPER(AmdDxExtMultidrawIndirect_MultiDrawIndirectCtx, 24)
+HRESULT __thiscall AmdDxExtMultidrawIndirect_MultiDrawIndirectCtx(IAmdDxExtMultidrawIndirect *iface, unsigned int draw_count, ID3D11Buffer *buffer,
+                                                                  unsigned int byte_offset, unsigned int byte_stride, ID3D11DeviceContext *ctx)
+{
+    ID3D11VkExtContext *ext_context;
+    TRACE("%p %u %p %u %u %p\n", iface, draw_count, buffer, byte_offset, byte_stride, ctx);
+
+    if (FAILED(ID3D11DeviceContext_QueryInterface(ctx, &IID_ID3D11VkExtContext, (void **)&ext_context)))
+        return E_FAIL;
+
+    ID3D11VkExtContext_MultiDrawIndirect(ext_context, draw_count, buffer, byte_offset, byte_stride);
+    ID3D11VkExtContext_Release(ext_context);
+
+    return S_OK;
+}
+
+DEFINE_THISCALL_WRAPPER(AmdDxExtMultidrawIndirect_MultiDrawIndexedIndirectCtx, 24)
+HRESULT __thiscall AmdDxExtMultidrawIndirect_MultiDrawIndexedIndirectCtx(IAmdDxExtMultidrawIndirect *iface, unsigned int draw_count, ID3D11Buffer *buffer,
+                                                                         unsigned int byte_offset, unsigned int byte_stride, ID3D11DeviceContext *ctx)
+{
+    ID3D11VkExtContext *ext_context;
+    TRACE("%p %u %p %u %u %p\n", iface, draw_count, buffer, byte_offset, byte_stride, ctx);
+
+    if (FAILED(ID3D11DeviceContext_QueryInterface(ctx, &IID_ID3D11VkExtContext, (void **)&ext_context)))
+        return E_FAIL;
+
+    ID3D11VkExtContext_MultiDrawIndexedIndirect(ext_context, draw_count, buffer, byte_offset, byte_stride);
+    ID3D11VkExtContext_Release(ext_context);
+
+    return S_OK;
+}
+
+DEFINE_THISCALL_WRAPPER(AmdDxExtMultiDrawIndirect_MultiDrawIndirectCountCtx, 28)
+HRESULT __thiscall AmdDxExtMultiDrawIndirect_MultiDrawIndirectCountCtx(IAmdDxExtMultidrawIndirect *iface, ID3D11Buffer *buffer_for_count, unsigned int byte_offset_for_count,
+                                                                       ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride, ID3D11DeviceContext *ctx)
+{
+    AmdDxExt *This = impl_from_IAmdDxExtMultidrawIndirect(iface);
+    ID3D11VkExtContext *ext_context;
+    unsigned int max_draw_count;
+    TRACE("%p %p %u %p %u %u %p\n", iface, buffer_for_count, byte_offset_for_count, buffer, byte_offset, byte_stride, ctx);
+
+    if (!This->multi_draw_indirect_count) return E_FAIL;
+
+    max_draw_count = get_max_draw_count(buffer, byte_offset, byte_stride, sizeof(D3D11_DRAW_INDEXED_INSTANCED_INDIRECT_ARGS));
+
+    if (FAILED(ID3D11DeviceContext_QueryInterface(ctx, &IID_ID3D11VkExtContext, (void **)&ext_context)))
+        return E_FAIL;
+
+    ID3D11VkExtContext_MultiDrawIndirectCount(ext_context, max_draw_count, buffer_for_count, byte_offset_for_count, buffer, byte_offset, byte_stride);
+    ID3D11VkExtContext_Release(ext_context);
+
+    return S_OK;
+}
+
+DEFINE_THISCALL_WRAPPER(AmdDxExtMultiDrawIndirect_MultiDrawIndexedIndirectCountCtx, 28)
+HRESULT __thiscall AmdDxExtMultiDrawIndirect_MultiDrawIndexedIndirectCountCtx(IAmdDxExtMultidrawIndirect *iface, ID3D11Buffer *buffer_for_count, unsigned int byte_offset_for_count,
+                                                                              ID3D11Buffer *buffer, unsigned int byte_offset, unsigned int byte_stride, ID3D11DeviceContext *ctx)
+{
+    AmdDxExt *This = impl_from_IAmdDxExtMultidrawIndirect(iface);
+    ID3D11VkExtContext *ext_context;
+    unsigned int max_draw_count;
+    TRACE("%p %p %u %p %u %u %p\n", iface, buffer_for_count, byte_offset_for_count, buffer, byte_offset, byte_stride, ctx);
+
+    if (!This->multi_draw_indirect_count) return E_FAIL;
+
+    max_draw_count = get_max_draw_count(buffer, byte_offset, byte_stride, sizeof(D3D11_DRAW_INDEXED_INSTANCED_INDIRECT_ARGS));
+
+    if (FAILED(ID3D11DeviceContext_QueryInterface(ctx, &IID_ID3D11VkExtContext, (void **)&ext_context)))
+        return E_FAIL;
+
+    ID3D11VkExtContext_MultiDrawIndexedIndirectCount(ext_context, max_draw_count, buffer_for_count, byte_offset_for_count, buffer, byte_offset, byte_stride);
+    ID3D11VkExtContext_Release(ext_context);
+
+    return S_OK;
+}
+
 static const IAmdDxExtUAVOverlapVtbl amddxext_uav_vtable =
 {
     THISCALL(AmdDxExtUAVOverlap_AddRef),
@@ -678,7 +818,9 @@ static const IAmdDxExtUAVOverlapVtbl amddxext_uav_vtable =
     THISCALL(AmdDxExtUAVOverlap_Destroy),
     THISCALL(AmdDxExtUAVOverlap_BeginUAVOverlap),
     THISCALL(AmdDxExtUAVOverlap_EndUAVOverlap),
-    THISCALL(AmdDxExtUAVOverlap_GetVersion)
+    THISCALL(AmdDxExtUAVOverlap_GetVersion),
+    THISCALL(AmdDxExtUAVOverlap_BeginUAVOverlapCtx),
+    THISCALL(AmdDxExtUAVOverlap_EndUAVOverlapCtx)
 };
 
 static const IAmdDxExtVtbl AmdDxExt_vtable =
@@ -713,7 +855,8 @@ static const IAmdDxExtDepthBoundsVtbl amddxext_depth_vtable =
     THISCALL(AmdDxExtDepthBounds_Release),
     THISCALL(AmdDxExtDepthBounds_Destroy),
     THISCALL(AmdDxExtDepthBounds_SetDepthBounds),
-    THISCALL(AmdDxExtDepthBounds_GetVersion)
+    THISCALL(AmdDxExtDepthBounds_GetVersion),
+    THISCALL(AmdDxExtDepthBounds_SetDepthBoundsCtx),
 };
 
 static const IAmdDxExtMultidrawIndirectVtbl amddxext_multidraw_vtable =
@@ -724,8 +867,12 @@ static const IAmdDxExtMultidrawIndirectVtbl amddxext_multidraw_vtable =
     THISCALL(AmdDxExtMultidrawIndirect_GetVersion),
     THISCALL(AmdDxExtMultidrawIndirect_MultiDrawIndirect),
     THISCALL(AmdDxExtMultidrawIndirect_MultiDrawIndexedIndirect),
-    THISCALL(AmdDxextMultiDrawIndirect_MultiDrawIndirectCount),
-    THISCALL(AmdDxextMultiDrawIndirect_MultiDrawIndexedIndirectCount)
+    THISCALL(AmdDxExtMultiDrawIndirect_MultiDrawIndirectCount),
+    THISCALL(AmdDxExtMultiDrawIndirect_MultiDrawIndexedIndirectCount),
+    THISCALL(AmdDxExtMultidrawIndirect_MultiDrawIndirectCtx),
+    THISCALL(AmdDxExtMultidrawIndirect_MultiDrawIndexedIndirectCtx),
+    THISCALL(AmdDxExtMultiDrawIndirect_MultiDrawIndirectCountCtx),
+    THISCALL(AmdDxExtMultiDrawIndirect_MultiDrawIndexedIndirectCountCtx),
 };
 
 HRESULT CDECL AmdDxExtCreate11(ID3D11Device *device, IAmdDxExt **ext)
-- 
2.55.0

